File: solution/day23/src/main/java/org/craftedsw/tripservicekata/trip/TripService.java
package org.craftedsw.tripservicekata.trip;

import io.vavr.collection.Seq;
import io.vavr.control.Try;
import org.craftedsw.tripservicekata.exception.UserNotLoggedInException;
import org.craftedsw.tripservicekata.user.User;

import java.util.function.Supplier;

import static io.vavr.collection.List.empty;
import static io.vavr.control.Try.failure;
import static io.vavr.control.Try.success;

public class TripService {
    private static final Seq<Trip> NO_TRIPS = empty();
    private final TripRepository repository;

    public TripService(TripRepository repository) {
        this.repository = repository;
    }

    public Try<Seq<Trip>> retrieveFriendTrips(User user, User loggedUser) {
        return checkUser(loggedUser, () -> retrieveFriendsSafely(user, loggedUser));
    }

    private Seq<Trip> retrieveFriendsSafely(User user, User loggedUser) {
        return user.isFriendWith(loggedUser)
                ? findTripsByUser(user)
                : NO_TRIPS;
    }

    private Try<Seq<Trip>> checkUser(User loggedUser,
                                     Supplier<Seq<Trip>> continueWith) {
        return (loggedUser != null)
                ? success(continueWith.get())
                : failure(new UserNotLoggedInException());
    }

    protected Seq<Trip> findTripsByUser(User user) {
        return repository.findTripsByUser(user);
    }
}

File: solution/kotlin/day23/src/main/kotlin/org/craftedsw/tripservicekata/trip/TripService.kt
package org.craftedsw.tripservicekata.trip

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException
import org.craftedsw.tripservicekata.user.User

class TripService(private val tripRepository: TripRepository) {
    private val NO_TRIPS = emptyList<Trip>();

    fun retrieveFriendTrips(user: User, loggedUser: User?): Result<List<Trip>> {
        return checkUser(loggedUser) { retrieveFriendsSafely(user, loggedUser!!) }
    }

    private fun retrieveFriendsSafely(user: User, loggedUser: User): List<Trip> =
        if (user.isFriendWith(loggedUser)) findTripsByUser(user)
        else NO_TRIPS

    private fun checkUser(loggedUser: User?, continueWith: () -> List<Trip>): Result<List<Trip>> =
        if (loggedUser != null) Result.success(continueWith())
        else Result.failure(UserNotLoggedInException())

    private fun findTripsByUser(user: User): List<Trip> = tripRepository.findTripsByUser(user)
}